### PR TITLE
Task/DES-1988: Save to project metadata

### PR DESCRIPTION
**Pull request by duckonomy**
## Overview: ##
Implements file save in client and adds post to project metadata on saving a map.
Requires #69 to be merged. 

## PR Status: ##

* [X] Ready.
* [ ] Work in Progress.
* [ ] Hold.

## Related Jira tickets: ##

* [DES-1988](https://jira.tacc.utexas.edu/browse/DES-1988)

## Testing Steps: ##
1. Test with tacc-cloud/geoapi#68 to ensure that saving works through only the client.
2. Ensure that the metadata is updated for DesignSafe.

## Notes: ##
**TODO**
* [ ] Merge #69

### CODE REVIEW

File: src/app/services/agave-systems.service.ts [hunk 1]
```diff
@@ -51,4 +51,67 @@ export class AgaveSystemsService {
...
+
+    const fileType = "plain/text";
+    const tmp = new Blob([data], {type: fileType});
+    const date = new Date();
+    const file = new File([tmp], fileName + ".hazmapper", {lastModified: date.valueOf()});
+
+    const form: FormData = new FormData;
+    form.append("fileToUpload", file)
+
+    this.http.post(this.envService.designSafeUrl + `files/v2/media/system/${systemID}${path}`, form).subscribe(resp => {
```
[the remark this answers is not in the document]
Comment (nathanfranklin) [reply src/app/services/agave-systems.service.ts:103]: also should there be a `/` between systemID and path in that url?

**Review by nathanfranklin — COMMENTED** [review]
Looks good 👍  💯 

I left some comments to consider.  The more complicated error handling ones we should just spin off into follow on ticket for some later time.


-----------
Note, I edited `modal-create-project.component.ts` locally to include a call to `this.projectsService.exportProject`  in `createRapidProject` to get things to work

File: src/app/services/agave-systems.service.ts [hunk 2]
```diff
@@ -51,4 +51,67 @@ export class AgaveSystemsService {
...
+      };
+
+      const headers = new HttpHeaders()
+        .set('X-Requested-With', 'XMLHttpRequest');
+
+      this.http.post<any>(this.envService.designSafeUrl + `projects-staging/v2/${designsafeUUID}/`, payload, {headers})
+        .subscribe( resp => {
+          console.log(resp);
+        }, error => {
+          console.log(error);
```
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/services/agave-systems.service.ts:85]: I'll try to handle this on this PR.

File: src/app/services/projects.service.ts [hunk 3]
```diff
@@ -108,11 +108,18 @@ export class ProjectsService {
       file_name,
       link
     };
+    this.agaveSystemsService.saveDSFile(system_id, path, file_name, project);
+    if (system_id.includes('project')) {
```
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/services/projects.service.ts:112]: That sounds good!

File: src/app/services/agave-systems.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/services/agave-systems.service.ts:103]: > 

The `path` already has a `/` included.
Would it be safe to include an extra `/`?

File: src/app/services/agave-systems.service.ts [hunk 4]
```diff
@@ -51,4 +51,67 @@ export class AgaveSystemsService {
     }
     return projects;
   }
+
+  updateDSProjectInformation(designsafeUUID: string, path: string, project: Project, operation: string) {
```
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/services/agave-systems.service.ts:55]: Gotcha!

File: src/app/services/agave-systems.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nathanfranklin) [reply src/app/services/agave-systems.service.ts:103]: > The path already has a / included.
> Would it be safe to include an extra /

now i'm not certain. probably best to just ignore my comment.  i think i saw made this but i had goofed things up when editing createRapidProject which was probably the problem

**Comment by duckonomy** [comment]
@nathanfranklin Made the adjustments :)
I think there is a bug for the json file save (the DS side has trouble parsing it).. I'll try to investigate locally.
Both `ng-tapis` and making a direct request gave me the same results.

**Pull request merged**